### PR TITLE
[DRAFT]: Refactor of diffusion samplers

### PR DESCRIPTION
**Pull request by CharlelieLrt**
# PhysicsNeMo Pull Request

## Description

**Note:** this PR is currently a draft and is open for feedback and suggestions. It will not be merged in its current form and might be broken down into smaller PRs.

### Objectives

**Overarching objective:** refactor and improve all utilities and and models related to diffusion in order to consolidate these components into a _framework_. More precisely, these diffusion components should:

- Support out of the box all current diffusion use-cases (CorrDiff, cBottle, StormCast, etc.)
- Be composable
- Be extendable
- Be well-documented

**PR objective:** focuses on the EDM samplers (`stochastic_sampler` and `deterministic_sampler`). For these samplers, the objective is to be:

1. Agnostic to the diffusion model
2. Agnostic to the modality of the latent state `x`
3. Agnostic to the modality of the conditioning (for conditional diffusion models)
4. Support a large range of guidance for plug-and-play generation (e.g. DPS)
5. Support multiple implementation of multi-diffusion

### Solutions

0. Refactored the `stochastic_sampler` functional interface into an object-oriented interface to facilitate future extensibility.

1. _Model agnostic:_ relies on a callback whose signature is assumed invariant. To be able to satisfy this invariant constraint we also provide a surface adapter (i.e. a thin wrapper) that modifies the signature of any given `Module` to ensure compliance.

2. _Latent state agnostic:_ simple refactors to avoid unnecessary assumptions on the shape of the latent state.

3. _Conditioning agnostic:_ all conditioning variables are packed into a dict of tensors. The sampler never accesses the conditioning as the model is responsible for handling the conditioning ops.

4. _Plug-and-play guidance:_ relies on callbacks passed to the sampler. Introduces a guidance API to facilitate creating these callbacks, and ensure compliance with the sampler requirements. For now two types of guidance are provided (model-based guidance for DPS, and data consistency guidance for inpainting/outpainting/channel infilling, etc.)

5. _Multi-diffusion:_ TBD. For now the plan is to defer multi-diffusion ops to a dedicated model wrapper.

### Remaining items

_- Multi-diffusion:_ in their current implementation, the samplers use a `patching` object to extract patches from the latent state `x`; it also calls methods from the model to extract the global positional embedding of these patches. These strong assumptions on the model APIs are not compatible with the objective (1) above. A better solution might be to defer all multi-diffusion ops to a model wrapper that extract patches, get the positional embeddings ,etc...

_- Guidance:_ only support pre-defined guidance, no mechanism to allow user-defined guidance. Extend the range of available off-the-shelf guidances.

_- Model-based guidance:_ only support a model that processes batches, and that is implemented in pytorch and compatible with `torch.autograd`.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

### CODE REVIEW

File: physicsnemo/experimental/utils/diffusion/guidance.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        # NOTE: for now only Tweedie's formula to estimate clean state x_0
+        y_x0: torch.Tensor = self.guide_model(x_0_hat)  # (*_y,)
+        if y_x0.shape != y.shape:
+            raise ValueError(
+                f"Expected 'guide_model' output and y to have same shape, "
+                f"but got {y_x0.shape} and {y.shape}"
+            )
+        err1 = torch.abs((y - y_x0)) ** self.norm_ord  # (*_y,)
+
+        # Compute log-likelihood p(y|x_0_hat)
```
Comment (ManshaP) [inline physicsnemo/experimental/utils/diffusion/guidance.py:173]: This is relatively specific to DPS, I believe. Other model-based guidance approaches may use a different parameterization of the time-dependent variance (rather than with gamma), or a different loss altogether (cBottle TC uses BinaryCrossEntropy

File: physicsnemo/experimental/utils/diffusion/samplers.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+            **model_kwargs,
+        )
+
+        # Guidance (e.g. posterior sampling, etc...)
+        guidance_sum = 0.0
+        if guidances:
+            for guidance, guidance_args, guidance_kwargs in zip(
+                guidances, guidances_args, guidances_kwargs
+            ):
+                if isinstance(guidance, ModelBasedGuidance):
```
Comment (ManshaP) [inline physicsnemo/experimental/utils/diffusion/samplers.py]: There is a fundamental difference between kinds of guidance that use the clean and the noisy state (e.g. cBottle TC) so maybe worth making a separate class for guidance that does not require the clean state?

File: physicsnemo/experimental/utils/diffusion/samplers.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+                B,
+            ),
+            cond,
+            *model_args,
+            **model_kwargs,
+        )
+
+        # Guidance (e.g. posterior sampling, etc...)
+        guidance_sum = 0.0
+        if guidances:
```
Comment (ManshaP) [inline physicsnemo/experimental/utils/diffusion/samplers.py]: I like this modularity/allowing for multiple guidance terms!

File: physicsnemo/experimental/utils/diffusion/samplers.py [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+            ).to(x.device)
+
+            # Move conditioning to the device
+            for key, value in cond.items():
+                cond[key] = value.to(x.device)
+
+            # Activate gradient computation if needed for guidance
+            with torch.set_grad_enabled(req_grad):
+                if req_grad:
+                    x_hat_in = x_hat.clone().detach().requires_grad_(True)
```
Comment (juliusberner) [inline physicsnemo/experimental/utils/diffusion/samplers.py:355]: nit: first detaching and then cloning is more efficient.

File: physicsnemo/experimental/utils/diffusion/samplers.py [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+            ).to(memory_format=torch.channels_last)
+
+            # Call the sampler function
+            x_0: torch.Tensor = sampler_fn(x_T, cond, **sampler_kwargs)
+
+            x_generated.append(x_0)
+    return torch.cat(x_generated)
+
+
+class EDMStochasticSampler:
```
Comment (juliusberner) [inline physicsnemo/experimental/utils/diffusion/samplers.py:118]: @CharlelieLrt do you plan to decouple the sampler from the choice of time discretization? E.g., we could use this Heun sampler with other time steps than the ones proposed in the EDM paper.

File: physicsnemo/experimental/utils/diffusion/guidance.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Callable, Literal
+
+import torch
+
+
+class ModelBasedGuidance:
```
Comment (juliusberner) [inline physicsnemo/experimental/utils/diffusion/guidance.py:22]: I would maybe call this DPSGuidance (to be more precise, this implementation also assumes Gaussian noise model), since there are several other guidance methods; see, e.g., https://arxiv.org/pdf/2503.11043 for an overview.

File: physicsnemo/experimental/utils/diffusion/adapter.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+          method that the noise level ``sigma`` should be mapped to.
+        - Third element: a dictionary mapping keys in the `cond` dictionary
+          to parameter names in the original model's forward method.
+
+    Forward
+    -------
+    x : torch.Tensor
+        The latent state of the diffusion model, typically of shape
+        :math:`(B, *)`.
+    sigma : torch.Tensor
```
Comment (juliusberner) [inline physicsnemo/experimental/utils/diffusion/adapter.py:56]: I think more generally the input to the model is `t` (which just coincides with `sigma` for the VE schedule in the EDM formulation).

File: physicsnemo/experimental/utils/diffusion/guidance.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+    r"""
+    Guidance function for `Diffusion Posterior Sampling (DPS)
+    <https://arxiv.org/abs/2209.14687>`_ based on a generic user-defined model
+    (possibly non-linear).
+    An instance of ``ModelBasedGuidance`` is a callable object that returns
+    the likelihood score :math:`\nabla_{\mathbf{x}_t} \log
+    p(\mathbf{y}|\mathbf{x}_t)`, where :math:`\mathbf{y}` is some conditioniong
```
Comment (simonbyrne) [inline physicsnemo/experimental/utils/diffusion/guidance.py:29]: typo
```suggestion
    p(\mathbf{y}|\mathbf{x}_t)`, where :math:`\mathbf{y}` is some conditioning
```

**Comment by CharlelieLrt** [comment]
Closing because changes were included in #1363 .
